synapse/storage/data_stores/main/push_rule.py

```diff
@@ -77,6 +80,15 @@ class PushRulesWorkerStore(
     def __init__(self, database: Database, db_conn, hs):
         super(PushRulesWorkerStore, self).__init__(database, db_conn, hs)
 
+        if hs.config.worker.worker_app is None:
+            self._push_rules_stream_id_gen = ChainedIdGenerator(
+                self._stream_id_gen, db_conn, "push_rules_stream", "stream_id"
+            )  # type: Union[ChainedIdGenerator, SlavedIdTracker]
```

Comment:
Yeah, I briefly thought about it then screamed a bit. We could make a protocol that encompasses both but then I'm not sure that is that helpful if half the functions are stubbed out in both. In general the typing here is a bit of a mess and I think we should do something properly about it

Reply:
(I'm happy to add a protocol here, just not sure its usefulness in practice)

Reply:
fair enough. I do think there needs to be some sort of class here, if only because we'll need it to get mypy working properly. However, I'm happy for it to be punted for now.

Reply:
Agreed, though I think we'll need to do some refactoring of some form to make the types make sense, possibly.